module/VuFind/src/VuFind/Db/Service/LoginTokenService.php

```diff
@@ -147,8 +183,29 @@ public function deleteByUser(UserEntityInterface|int $userOrId): void
      */
     public function getByUser(UserEntityInterface|int $userOrId, bool $grouped = true): array
     {
-        $userId = is_int($userOrId) ? $userOrId : $userOrId->getId();
-        return $this->getDbTable('LoginToken')->getByUserId($userId, $grouped);
+        $user = $this->getDoctrineReference(User::class, $userOrId);
+        $dql = 'SELECT lt '
+            . 'FROM ' . $this->getEntityClass(LoginTokenEntityInterface::class) . ' lt '
+            . 'WHERE lt.user = :user '
+            . 'ORDER BY lt.lastLogin DESC';
+
+        if ($grouped) {
+            // Modify the DQL for grouping logic
+            $dql = 'SELECT lt '
+                . 'FROM ' . $this->getEntityClass(LoginTokenEntityInterface::class) . ' lt '
+                . 'WHERE lt.user = :user AND lt.lastLogin = ('
+                . '    SELECT MAX(subLt.lastLogin) '
+                . '    FROM ' . $this->getEntityClass(LoginTokenEntityInterface::class) . ' subLt '
+                . '    WHERE subLt.user = :user AND subLt.series = lt.series AND subLt.browser = lt.browser '
+                . '        AND subLt.platform = lt.platform AND subLt.expires = lt.expires '
+                . ') '
+                . 'ORDER BY lt.lastLogin DESC';
+        }
```

Comment:
I think an if/else here might be slightly more efficient so we don't assign `$dql` twice in the `$grouped` case.

```suggestion
        if ($grouped) {
            // Use different DQL for grouping logic
            $dql = 'SELECT lt '
                . 'FROM ' . $this->getEntityClass(LoginTokenEntityInterface::class) . ' lt '
                . 'WHERE lt.user = :user AND lt.lastLogin = ('
                . '    SELECT MAX(subLt.lastLogin) '
                . '    FROM ' . $this->getEntityClass(LoginTokenEntityInterface::class) . ' subLt '
                . '    WHERE subLt.user = :user AND subLt.series = lt.series AND subLt.browser = lt.browser '
                . '        AND subLt.platform = lt.platform AND subLt.expires = lt.expires '
                . ') '
                . 'ORDER BY lt.lastLogin DESC';
        } else {
            $dql = 'SELECT lt '
                . 'FROM ' . $this->getEntityClass(LoginTokenEntityInterface::class) . ' lt '
                . 'WHERE lt.user = :user '
                . 'ORDER BY lt.lastLogin DESC';
        }
```